flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTriggeringTest.java

```diff
@@ -283,31 +300,494 @@ public void testStopPeriodicScheduler() throws Exception {
+		try {
+			onCompletionPromise1.get();
+			fail("Should not reach here");
+		} catch (ExecutionException e) {
+			final Optional<CheckpointException> checkpointExceptionOptional =
+				ExceptionUtils.findThrowable(e, CheckpointException.class);
+			assertTrue(checkpointExceptionOptional.isPresent());
+			assertEquals(CheckpointFailureReason.TRIGGER_CHECKPOINT_FAILURE,
+				checkpointExceptionOptional.get().getCheckpointFailureReason());
+		}
+		assertFalse(coord.isTriggering());
+		assertEquals(0, coord.getTriggerRequestQueue().size());
+
+		final CompletableFuture<CompletedCheckpoint> onCompletionPromise2 = coord.triggerCheckpoint(
+			System.currentTimeMillis(),
+			CheckpointProperties.forCheckpoint(CheckpointRetentionPolicy.NEVER_RETAIN_AFTER_TERMINATION),
+			null,
+			true,
+			false);
+		assertTrue(coord.isTriggering());
+		manuallyTriggeredScheduledExecutor.triggerAll();
+		assertFalse(onCompletionPromise2.isCompletedExceptionally());
+		assertFalse(coord.isTriggering());
+		assertEquals(0, coord.getTriggerRequestQueue().size());
+	}
+
+	@Test
+	public void testTriggerCheckpointSnapshotMasterHookFailed() throws Exception {
+		// create some mock Execution vertices that receive the checkpoint trigger messages
+		final ExecutionAttemptID attemptID1 = new ExecutionAttemptID();
+		final AtomicInteger taskManagerCheckpointTriggeredTimes = new AtomicInteger(0);
+		final SimpleAckingTaskManagerGateway.CheckpointConsumer checkpointConsumer =
+			(executionAttemptID,
+			jobId, checkpointId,
+			timestamp,
+			checkpointOptions,
+			advanceToEndOfEventTime) -> taskManagerCheckpointTriggeredTimes.incrementAndGet();
+		ExecutionVertex vertex1 = mockExecutionVertex(attemptID1, checkpointConsumer);
+
+		// set up the coordinator and validate the initial state
+		CheckpointCoordinatorConfiguration chkConfig = new CheckpointCoordinatorConfiguration(
+			600000,
+			600000,
+			0,
+			Integer.MAX_VALUE,
+			CheckpointRetentionPolicy.NEVER_RETAIN_AFTER_TERMINATION,
+			true,
+			false,
+			0);
+		CheckpointCoordinator coord = new CheckpointCoordinator(
+			new JobID(),
+			chkConfig,
+			new ExecutionVertex[] { vertex1 },
+			new ExecutionVertex[] { vertex1 },
+			new ExecutionVertex[] { vertex1 },
+			new StandaloneCheckpointIDCounter(),
+			new StandaloneCompletedCheckpointStore(1),
+			new MemoryStateBackend(),
+			Executors.directExecutor(),
+			manuallyTriggeredScheduledExecutor,
+			SharedStateRegistry.DEFAULT_FACTORY,
+			failureManager);
+
+		final CompletableFuture<String> masterHookCheckpointFuture = new CompletableFuture<>();
+		coord.addMasterHook(new TestingMasterHook(masterHookCheckpointFuture));
+		coord.startCheckpointScheduler();
+		// Periodic
+		final CompletableFuture<CompletedCheckpoint> onCompletionPromise = coord.triggerCheckpoint(
+			System.currentTimeMillis(),
+			CheckpointProperties.forCheckpoint(CheckpointRetentionPolicy.NEVER_RETAIN_AFTER_TERMINATION),
+			null,
+			true,
+			false);
```

Comment:
maybe extract to a named method, like `onCompletionPromise = triggerPeriodicCheckpoint(coordinator)`? This way you would deduplicate some code and could remove the `// Periodic` comment (`selfDocumentingCode` > `comment`)